hazelcast/src/main/java/com/hazelcast/sql/impl/SqlInternalService.java

```diff
@@ -195,4 +195,27 @@ public QueryStateRegistry getStateRegistry() {
     public QueryOperationHandlerImpl getOperationHandler() {
         return operationHandler;
     }
+
+    private void prepareParameters(Plan plan, List<Object> params) {
+        assert params != null;
+        QueryParameterMetadata parameterMetadata = plan.getParameterMetadata();
+
+        int parameterCount = parameterMetadata.getParameterCount();
+        if (parameterCount != params.size()) {
+            throw QueryException.error("Unexpected parameter count: expected " + parameterCount + ", got " + params.size());
+        }
+
+        for (int i = 0; i < params.size(); ++i) {
+            Object value = params.get(i);
+            if (value == null) {
+                continue;
+            }
+
+            Converter valueConverter = Converters.getConverter(value.getClass());
+            Converter typeConverter = parameterMetadata.getParameterType(i).getConverter();
+            value = typeConverter.convertToSelf(valueConverter, value);
```

Comment:
What if conversion fails here? Users should receive a proper exception that a parameter on the position N of type T1 cannot be converted to a type T2. Looks like this is not the case.

Reply:
fixed